src/TelemetryChannels/ServerTelemetryChannel/Shared/Implementation/SamplingPercentageEstimatorTelemetryProcessor.cs

```diff
@@ -0,0 +1,244 @@
+            AdaptiveSamplingPercentageEvaluatedCallback callback, 
+            ITelemetryProcessor next)
+        {
+            if (settings == null)
+            {
+                throw new ArgumentNullException("settings");
+            }
+
+            if (next == null)
+            {
+                throw new ArgumentNullException("next");
+            }
+
+            this.evaluationCallback = callback;
+            this.settings = settings;
+            this.next = next;
+
+            this.currenSamplingRate = settings.EffectiveInitialSamplingRate;
+
+            this.itemCount = new ExponentialMovingAverageCounter(settings.EffectiveMovingAverageRatio);
+
+            this.samplingPercentageLastChangeDateTime = DateTimeOffset.UtcNow;
+
+            // set evaluation interval to default value if it is negative or zero
+            this.evaluationInterval = this.settings.EffectiveEvaluationInterval;
+
+            // set up timer to run math to estimate sampling percentage
+            this.evaluationTimer = new Timer(
+                this.EstimateSamplingPercentage, 
+                null,
+                this.evaluationInterval,
+                this.evaluationInterval);
+        }
+
+        /// <summary>
+        /// Processes telemetry item.
+        /// </summary>
+        /// <param name="item">Telemetry item to process.</param>
+        public void Process(ITelemetry item)
+        {
+            // increment post-samplin telemetry item counter
+            this.itemCount.Increment();
+
+            // continue processing telemetry item with the next telemetry processor
+            this.next.Process(item);
+        }
+
+        /// <summary>
+        /// Disposes the object.
+        /// </summary>
+        public void Dispose()
+        {
+            if (this.evaluationTimer != null)
+            {
+                this.evaluationTimer.Dispose();
+                this.evaluationTimer = null;
+            }
+        }
+
+        /// <summary>
+        /// Checks to see if exponential moving average has changed.
+        /// </summary>
+        /// <param name="running">Currently running value of moving average.</param>
+        /// <param name="current">Value set in the algorithm parameters.</param>
+        /// <returns>True if moving average value changed.</returns>
+        private static bool MovingAverageCoefficientChanged(double running, double current)
+        {
+            const double Precision = 1E-12;
+
+            return (running < current - Precision) || (running > current + Precision);
+        }
+
+        /// <summary>
+        /// Callback for sampling percentage evaluation timer.
+        /// </summary>
+        /// <param name="state">Timer state.</param>
+        private void EstimateSamplingPercentage(object state)
+        {
+            // get observed after-sampling eps
+            double observedEps = this.itemCount.StartNewInterval() / this.evaluationInterval.TotalSeconds;
```

Comment:
I was thinking about it.... I think it is fine as is. The interval would be in seconds usually and timer precision (plus overhead of callback execution) is few ms tops. So, it will be well within approximation boundaries...